# 1. Creat library and import dataframe

In [4]:
import pandas as pd
import numpy as np
import os

In [6]:
# Create Path
path = r'/Users/nhungphamthi/Documents/data analytis/Achievement 4 Project'

In [8]:
# Import ords_prods_merge dataframe 
df_ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged_derived.pkl'))

In [12]:
df_ords_prods_merge.shape

(32404859, 18)

# 2. Caculate aggregated mean for “order_number” column grouped by “department_id” for Entire Dataframe

In [10]:
# Group 'order_number'  by 'department_id', then Calculating Mean using agg() Function
df_ords_prods_merge.groupby('department_id').agg({'order_number' : ['mean']})

order_number
                      mean
department_id             
1                15.457838
2                17.277920
3                17.170395
4                17.811403
5                15.215751
6                16.439806
7                17.225802
8                15.340650
9                15.895474
10               20.197148
11               16.170638
12               15.887671
13               16.583536
14               16.773669
15               16.165037
16               17.665606
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

# 3. How do the results for the entire dataframe differ from those of the subset?

Comparing to the results from the subset, the results from the entire dataframe includes all'department_id' values and therefore the mean values for the department ids present are different. Some have decreased while others increased.

# 4. Create "Loyalty flag" for existing customers using the transform() and loc() functions

In [12]:
# Aggregating Data with transform() Function
df_ords_prods_merge['max_order'] = df_ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)

/var/folders/6g/n887dsw94zx88dt3dsw_svcr0000gn/T/ipykernel_27744/3473307984.py:2: FutureWarning: The provided callable <function max at 0x10e4ff1a0> is currently using SeriesGroupBy.max. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "max" instead.
  df_ords_prods_merge['max_order'] = df_ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)


In [52]:
# Checking the Output and 'loyalty_flag' found.
df_ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_the_week  order_hour_of_day  \
0   2539329        1             1                       2                  8   
1   2539329        1             1                       2                  8   
2   2539329        1             1                       2                  8   
3   2539329        1             1                       2                  8   
4   2539329        1             1                       2                  8   

   days_since_prior_order  new_customer  product_id  add_to_cart_order  \
0                     NaN          True         196                  1   
1                     NaN          True       14084                  2   
2                     NaN          True       12427                  3   
3                     NaN          True       26088                  4   
4                     NaN          True       26405                  5   

   reordered  ...    price_range_loc     busiest_day    busiest_days  \
0          0  ...  Mid-range product  Regularly busy  Regularly busy   
1          0  ...  Mid-range product  Regularly busy  Regularly busy   
2          0  ...  Low-range product  Regularly busy  Regularly busy   
3          0  ...  Low-range product  Regularly busy  Regularly busy   
4          0  ...  Low-range product  Regularly busy  Regularly busy   

   busiest_period_of_day max_order  loyalty_flag avg_product_price  \
0         Average orders        10  New customer          6.367797   
1         Average orders        10  New customer          6.367797   
2         Average orders        10  New customer          6.367797   
3         Average orders        10  New customer          6.367797   
4         Average orders        10  New customer          6.367797   

  spending_flag  median_days_between_orders   order_frequency_flag  
0   Low spender                        20.5  Non-frequent customer  
1   Low spender                        20.5  Non-frequent customer  
2   Low spender                        20.5  Non-frequent customer  
3   Low spender                        20.5  Non-frequent customer  
4   Low spender                        20.5  Non-frequent customer  

[5 rows x 24 columns]

In [16]:
# Creating a 'Loyalty' Label Flag using loc Function
df_ords_prods_merge.loc[df_ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'
df_ords_prods_merge.loc[(df_ords_prods_merge['max_order'] <= 40) & (df_ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'
df_ords_prods_merge.loc[df_ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [18]:
# Examining Value Counts on 'loyalty_flag'
df_ords_prods_merge['loyalty_flag'].value_counts(dropna = False)

loyalty_flag
Regular customer    15876776
Loyal customer      10284093
New customer         6243990
Name: count, dtype: int64

In [55]:
# Checking the Updated Output
df_ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_the_week  order_hour_of_day  \
0   2539329        1             1                       2                  8   
1   2539329        1             1                       2                  8   
2   2539329        1             1                       2                  8   
3   2539329        1             1                       2                  8   
4   2539329        1             1                       2                  8   

   days_since_prior_order  new_customer  product_id  add_to_cart_order  \
0                     NaN          True         196                  1   
1                     NaN          True       14084                  2   
2                     NaN          True       12427                  3   
3                     NaN          True       26088                  4   
4                     NaN          True       26405                  5   

   reordered  ...    price_range_loc     busiest_day    busiest_days  \
0          0  ...  Mid-range product  Regularly busy  Regularly busy   
1          0  ...  Mid-range product  Regularly busy  Regularly busy   
2          0  ...  Low-range product  Regularly busy  Regularly busy   
3          0  ...  Low-range product  Regularly busy  Regularly busy   
4          0  ...  Low-range product  Regularly busy  Regularly busy   

   busiest_period_of_day max_order  loyalty_flag avg_product_price  \
0         Average orders        10  New customer          6.367797   
1         Average orders        10  New customer          6.367797   
2         Average orders        10  New customer          6.367797   
3         Average orders        10  New customer          6.367797   
4         Average orders        10  New customer          6.367797   

  spending_flag  median_days_between_orders   order_frequency_flag  
0   Low spender                        20.5  Non-frequent customer  
1   Low spender                        20.5  Non-frequent customer  
2   Low spender                        20.5  Non-frequent customer  
3   Low spender                        20.5  Non-frequent customer  
4   Low spender                        20.5  Non-frequent customer  

[5 rows x 24 columns]

# 5. Analyze spending habits by "loyalty category"

In [75]:
# Determining Basic Statistics of Spending Habits by Loyalty Category
df_ords_prods_merge.groupby('loyalty_flag').agg({'prices' : ['mean', 'min', 'max', 'std', 'median']})

prices                            
                      mean  min   max       std median
loyalty_flag                                          
Loyal customer    7.773488  1.0  25.0  4.205886    7.4
New customer      7.801077  1.0  25.0  4.266452    7.4
Regular customer  7.798145  1.0  25.0  4.254540    7.4

##### Loyal customers seem to buy cheap product (7.77), then regular customers (7.79) and new customers (7.8). I guess that the loyal customers can get some discounts in the most or they are noticed the lowest price by email advertisements, etc. 

# 6. Create a "spending flag"

In [62]:
# Aggregating Data with transform() Function
df_ords_prods_merge['avg_product_price'] = df_ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)

/var/folders/6g/n887dsw94zx88dt3dsw_svcr0000gn/T/ipykernel_27744/1144848887.py:2: FutureWarning: The provided callable <function mean at 0x10e4ffba0> is currently using SeriesGroupBy.mean. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "mean" instead.
  df_ords_prods_merge['avg_product_price'] = df_ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)


In [69]:
# Checking Relevant Columns 
df_ords_prods_merge[['user_id', 'avg_product_price']].head(100)

user_id  avg_product_price
0         1           6.367797
1         1           6.367797
2         1           6.367797
3         1           6.367797
4         1           6.367797
..      ...                ...
95        2           7.515897
96        2           7.515897
97        2           7.515897
98        2           7.515897
99        2           7.515897

[100 rows x 2 columns]

In [31]:
# Creating a 'Spending' Label Flag using loc Function
df_ords_prods_merge.loc[df_ords_prods_merge['avg_product_price'] >= 10, 'spending_flag'] = 'High spender'
df_ords_prods_merge.loc[df_ords_prods_merge['avg_product_price'] < 10, 'spending_flag'] = 'Low spender'

In [33]:
# Examining Value Count on 'spending_flag' Column
df_ords_prods_merge['spending_flag'].value_counts(dropna = False)

spending_flag
Low spender     32285150
High spender      119709
Name: count, dtype: int64

In [71]:
# Checking Releavent Columns
df_ords_prods_merge[['user_id', 'avg_product_price', 'spending_flag']].head(100)

user_id  avg_product_price spending_flag
0         1           6.367797   Low spender
1         1           6.367797   Low spender
2         1           6.367797   Low spender
3         1           6.367797   Low spender
4         1           6.367797   Low spender
..      ...                ...           ...
95        2           7.515897   Low spender
96        2           7.515897   Low spender
97        2           7.515897   Low spender
98        2           7.515897   Low spender
99        2           7.515897   Low spender

[100 rows x 3 columns]

# 7. Create an "order frequency flag"

In [38]:
# Aggregating Data with transform() Function
df_ords_prods_merge['median_days_between_orders'] = df_ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform(np.median)

/var/folders/6g/n887dsw94zx88dt3dsw_svcr0000gn/T/ipykernel_27744/4152215344.py:2: FutureWarning: The provided callable <function median at 0x10e677380> is currently using SeriesGroupBy.median. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "median" instead.
  df_ords_prods_merge['median_days_between_orders'] = df_ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform(np.median)


In [40]:
# Checking Relevant Columns 
df_ords_prods_merge[['user_id', 'days_since_prior_order', 'median_days_between_orders']].head(15)

user_id  days_since_prior_order  median_days_between_orders
0         1                     NaN                        20.5
1         1                     NaN                        20.5
2         1                     NaN                        20.5
3         1                     NaN                        20.5
4         1                     NaN                        20.5
5         1                    15.0                        20.5
6         1                    15.0                        20.5
7         1                    15.0                        20.5
8         1                    15.0                        20.5
9         1                    15.0                        20.5
10        1                    15.0                        20.5
11        1                    21.0                        20.5
12        1                    21.0                        20.5
13        1                    21.0                        20.5
14        1                    21.0                        20.5

In [42]:
# Creating a 'Order Frequency' Label Flag using loc Function
df_ords_prods_merge.loc[df_ords_prods_merge['median_days_between_orders'] > 20, 'order_frequency_flag'] = 'Non-frequent customer'
df_ords_prods_merge.loc[(df_ords_prods_merge['median_days_between_orders'] <= 20) & (df_ords_prods_merge['median_days_between_orders'] > 10), 'order_frequency_flag'] = 'Regular customer'
df_ords_prods_merge.loc[df_ords_prods_merge['median_days_between_orders'] <= 10, 'order_frequency_flag'] = 'Frequent customer'

In [44]:
# Examining Value Counts on 'order_frequency_flag'
df_ords_prods_merge['order_frequency_flag'].value_counts(dropna = False)

order_frequency_flag
Frequent customer        21559853
Regular customer          7208564
Non-frequent customer     3636437
NaN                             5
Name: count, dtype: int64

In [46]:
# Checking Relevant Columns of Updated Dataframe
df_ords_prods_merge[['user_id', 'days_since_prior_order', 'median_days_between_orders', 'order_frequency_flag']].head(15)

user_id  days_since_prior_order  median_days_between_orders  \
0         1                     NaN                        20.5   
1         1                     NaN                        20.5   
2         1                     NaN                        20.5   
3         1                     NaN                        20.5   
4         1                     NaN                        20.5   
5         1                    15.0                        20.5   
6         1                    15.0                        20.5   
7         1                    15.0                        20.5   
8         1                    15.0                        20.5   
9         1                    15.0                        20.5   
10        1                    15.0                        20.5   
11        1                    21.0                        20.5   
12        1                    21.0                        20.5   
13        1                    21.0                        20.5   
14        1                    21.0                        20.5   

     order_frequency_flag  
0   Non-frequent customer  
1   Non-frequent customer  
2   Non-frequent customer  
3   Non-frequent customer  
4   Non-frequent customer  
5   Non-frequent customer  
6   Non-frequent customer  
7   Non-frequent customer  
8   Non-frequent customer  
9   Non-frequent customer  
10  Non-frequent customer  
11  Non-frequent customer  
12  Non-frequent customer  
13  Non-frequent customer  
14  Non-frequent customer

# 8. Export the dataframe in pickle format to the "Prepared Data" folder

In [88]:
df_ords_prods_merge.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'ordes_products_merged_Ex4.8.pkl'))